add cross accounts and region testing in LS docs (#1129)
localstack/docs: content/en/contributing/multi-account-region-testing.md

---

> title: "Multi account and region testing"

Author: Got a review to change this [here](https://github.com/localstack/docs/pull/1129#discussion_r1528032339) earlier. @HarshCasper any suggestions here?

> weight: 7

Author: I discussed this with @HarshCasper since I wasn't sure about this. @alexrashed do you think it should be 8 here?

> weight: 7

Reviewer: I don't have a strong opinion here, but I would say the parity testing guide is a bit more important (because basically _every_ PR needs to use it, while this guide is only interesting for code touching the inter-service-communication).
The parity guide has weight 7, the terraform testing has 10, so I think 8 would work well.
But again, no strong opinion :)

description: >
  How to test multi-account and multi-region compatibility.
---

> LocalStack supports multi accounts and regions compatibility. This document describes how to test your changes against cross accounts and regions.

Reviewer 2: Please prefer having one sentence on each line, this helps with diffs in future changes.

> We regularly run the circleci test jobs of the LocalStack community repository to check the compatibility of cross accounts against the changes. To achieve that, we have a [scheduled workflow](https://github.com/localstack/localstack/blob/master/.circleci/config.yml) on [LocalStack](https://github.com/localstack/localstack), which executes the tests with randomised account and region credenitals every night at 1:00am UTC.

Reviewer 2: ```suggestion
We regularly run the circleci test jobs of the LocalStack community repository to check the compatibility of cross accounts against the changes. To achieve that, we have a [scheduled workflow](https://github.com/localstack/localstack/blob/master/.circleci/config.yml) on [LocalStack](https://github.com/localstack/localstack), which executes the tests with randomised account and region credentials every night at 1:00am UTC.
```

## Manually trigger the workflow

To manually trigger the scheduled workflow in case you have the permissions, you can change set the value of `randomize-aws-credentials` to `true` in the [workflow](https://github.com/localstack/localstack/blob/master/.circleci/config.yml#L13). This will trigger the tests with randomised account and region credentials.

{{< alert title="Note">}}
Make sure to change the value back to `false` after the tests are completed. 

> {{< /alert >}}

Author: This is in general for anyone wants to create a commit similar to one here- [10478](https://github.com/localstack/localstack/pull/10478/commits/b455596ee60ed012a3be8c5ec94ce4f5823bd4a7). This is to trigger the pipeline to check the results for non-defaults creds when someone doesnt has access to webapp directly.

> {{< /alert >}}

Reviewer: Ah, right, I see. And in fact, I overlooked the actual issue here:
What I meant in a previous comment to change this section describing how to manually trigger the pipeline with the parameter enabled _in the UI of CircleCI_, not in the code.
When the section above does not describe a code change anymore, this note also becomes unnecessary.

> {{< /alert >}}

Author: So to conclude, we want this part to include the param enabling through the webapp and not the code. Right @alexrashed?

---

> title: "Multiple account and region Testing"

Reviewer 3: ```suggestion
title: "Multi account and region testing"
```

weight: 7
description: >
  How to test multi-account and multi-region compatibility.
aliases:

>   - /developer-guide/multi-account-region-testing/

Reviewer 3: Don't need to add alias

---

> LocalStack community supports multiple accounts and regions compatibility. This document describes how to test your changes against cross accounts and regions.

Reviewer 3: ```suggestion
LocalStack supports multi accounts and regions compatibility. This document describes how to test your changes against cross accounts and regions.
```

Author: Since only community repo is multi accounts compatible for now, does it make sense to explicitly highlight this here @HarshCasper?

Author: like localstack community instead of just localstack? 🤔

Reviewer 3: We wish to remove any pricing info, and just mention `LocalStack` as a product in itself. Hence we can remove the `community` keyword.

## Manually trigger the workflow

To manually trigger the scheduled workflow, you can change set the value of `randomize-aws-credentials` to `true` in the [workflow](https://github.com/localstack/localstack/blob/master/.circleci/config.yml#L13). This will trigger the tests with randomised account and region credentials.

{{< alert title="Note">}}
Make sure to change the value back to `false` after the tests are completed. 
{{< /alert >}}

## Test changes locally with non-default credentials

In order to test your changes on your machine for multi-accounts and region compatibility, i.e. for non-default values other than `000000000000` for account ID or `us-east-1` for region, set the following environment variables to any random non-default values, for example:  

- `TEST_AWS_ACCOUNT_ID=111111111111`
- `TEST_AWS_ACCESS_KEY_ID=111111111111`
- `TEST_AWS_REGION=us-west-1`
- `TEST_AWS_SECRET_ACCESS_KEY=test1`

> We can additionally prefer to create a commit eg: [da3f8d5](https://github.com/localstack/localstack/pull/9751/commits/da3f8d5f2328adb7c5c025722994fea4433c08ba) to test the pipeline for non-default credentials against your changes.

Reviewer 3: ```suggestion
You can additionally prefer to create a commit eg: [da3f8d5](https://github.com/localstack/localstack/pull/9751/commits/da3f8d5f2328adb7c5c025722994fea4433c08ba) to test the pipeline for non-default credentials against your changes.
```

---
title: "Multi-account and Multi-region Testing"
weight: 8
description: >
  How to test multi-account and multi-region compatibility.
---

> LocalStack supports multi accounts and regions compatibility. 

Reviewer: nit: I think we should stick to the terms combined with a dash: `multi-account`, `multi-region`, `cross-account`, `inter-service-communication`. What's your opinion? Are there any guidelines or best practices on this, @HarshCasper?
```suggestion
LocalStack supports multi-accounts and multi-regions compatibility.
```

> This document describes how to test your changes against cross accounts and regions.

Reviewer: ```suggestion
This document describes how to test your changes for compatibility with cross-account and cross-region requests.
```

> We regularly run the circleci test jobs of the LocalStack community repository to check the compatibility of cross accounts against the changes. To achieve that, we have a [scheduled workflow](https://github.com/localstack/localstack/blob/master/.circleci/config.yml) on [LocalStack](https://github.com/localstack/localstack), which executes the tests with randomised account and region credentials every night at 1:00am UTC.

Reviewer: I think this isn't really useful information here. Maybe we could move this down to the section on manually triggering the pipeline? Maybe rephrase the section below to "Test changes in CI with non-default credentials" (in contrast and right next to "Test changes locally with non-default credentials")?
Also this line contains multiple sentences (we try to have a single sentence each line, see best practices in the README).

> ## Tips to make your changes compatible with multi-accounts and regions

Reviewer: ```suggestion
## Tips to make your changes compatible with multi-accounts and multi-regions
```

> In order to make your changes compatible with multi-accounts and regions, you should follow the below tips:

Reviewer: ```suggestion
In order to make your changes compatible with multi-accounts and multi-regions, you should follow the below tips:
```

> - For cross accounts, in inter service communication for many integrations, you can specify a role, with which permissions the source service makes a request to the target service, to access other services resource. This role should be in source account. When writing an AWS validated test case, you need to properly configure IAM roles. 

Reviewer: ```suggestion
- For cross-accounts, in inter-service-communication for many integrations, you can specify a role, with which permissions the source service makes a request to the target service, to access another service's resource.
  This role should be in source account.
  When writing an AWS validated test case, you need to properly configure IAM roles.
```

    ```

- While having an inter service communication in cross-accounts, you can create the client using `with_assumed_role(…)`:

    For example:
    ```python
    result = self.connect_api_gateway_to_kinesis(
            client,
            api_name,
            stream_name,
            region_name,
            role_arn,
        )
    ```
    
- When there is no role specified, you should use the source arn conceptually if cross-account is allowed. 

    For example:
    This can be seen in a case where we [added](https://github.com/localstack/localstack/blob/ae31f63bb6d8254edc0c85a66e3c36cd0c7dc7b0/localstack/utils/aws/message_forwarding.py#L42) `account_id` to [send events to the target](https://github.com/localstack/localstack/blob/ae31f63bb6d8254edc0c85a66e3c36cd0c7dc7b0/localstack/utils/aws/message_forwarding.py#L31) service like sqs, sns, lambda, etc. 

- You should always refer the official AWS API docs and check how the the application is communicating to each other. 
    
    For example: 
    Firehose to s3 refer to [Offical AWS Docs](https://docs.aws.amazon.com/firehose/latest/dev/controlling-access.html#cross-account-delivery-s3).

## Manually trigger the workflow

> To manually trigger the scheduled workflow in case you have the permissions, you can change set the value of `randomize-aws-credentials` to `true` in the [workflow](https://github.com/localstack/localstack/blob/master/.circleci/config.yml#L13). This will trigger the tests with randomised account and region credentials.

Reviewer: Please describe here how you can trigger the workflow with the proper parameters in the UI, and not with a commit.
Also, this line contains multiple sentences.

## Test changes locally with non-default credentials

In order to test your changes on your machine for multi-accounts and region compatibility in case you have the permissions, i.e. for non-default values other than `000000000000` for account ID or `us-east-1` for region, set the following environment variables to any random non-default values, for example:  

- `TEST_AWS_ACCOUNT_ID=111111111111`
- `TEST_AWS_ACCESS_KEY_ID=111111111111`
- `TEST_AWS_REGION=us-west-1`
- `TEST_AWS_SECRET_ACCESS_KEY=test1`

> You can additionally prefer to create a commit eg: [da3f8d5](https://github.com/localstack/localstack/pull/9751/commits/da3f8d5f2328adb7c5c025722994fea4433c08ba) to test the pipeline for non-default credentials against your changes. Note that within the tests we must use `account_id`, `secondary_account_id`, `region_name`, `secondary_region_name` fixtures and avoid importing `localstack.constants.TEST_*` values. 

Reviewer: ```suggestion
You can additionally prefer to create a commit eg: [da3f8d5](https://github.com/localstack/localstack/pull/9751/commits/da3f8d5f2328adb7c5c025722994fea4433c08ba) to test the pipeline for non-default credentials against your changes.
Note that within the tests we must use `account_id`, `secondary_account_id`, `region_name`, `secondary_region_name` fixtures and avoid importing `localstack.constants.TEST_*` values. 
```

---
title: "Multi-account and Multi-region Testing"
weight: 8
description: >
  How to test multi-account and multi-region compatibility.
---

LocalStack supports multi-account and multi-region compatibility. 
This document describes how to test your changes for compatibility with cross-account and cross-region requests.

## Tips to make your changes compatible with multi-account and multi-region

In order to make your changes compatible with multi-account and multi-region, you should follow the below tips:

- For cross-account, in inter-service-communication for many integrations, you can specify a role, with which permissions the source service makes a request to the target service, to access another service's resource. 

> This role should be in source account. 

Reviewer: nit:
```suggestion
This role should be in the source account. 
```

When writing an AWS validated test case, you need to properly configure IAM roles.

    For example: 

>     In the test case `test_api_gateway_kinesis_integration` we have specified a [role](https://github.com/localstack/localstack/blob/ae31f63bb6d8254edc0c85a66e3c36cd0c7dc7b0/tests/aws/services/apigateway/test_apigateway_basic.py#L2017-L2022) in the apigateway which has permissions to access the target kinesis account.

Reviewer: ```suggestion
    In the test case [`test_api_gateway_kinesis_integration`](https://github.com/localstack/localstack/blob/0c2d8a103a9a06ade4ef1c18eacf5888d2b8adcf/tests/aws/services/apigateway/test_apigateway_basic.py#L2047) we have specified a [role](https://github.com/localstack/localstack/blob/ae31f63bb6d8254edc0c85a66e3c36cd0c7dc7b0/tests/aws/services/apigateway/test_apigateway_basic.py#L2017-L2022) which has permissions to access the target kinesis account.
```

    For example:
    ```python
    result = self.connect_api_gateway_to_kinesis(
            client,
            api_name,
            stream_name,
            region_name,
            role_arn,
        )
    ```
    
- When there is no role specified, you should use the source arn conceptually if cross-account is allowed. 

    For example:
    This can be seen in a case where we [added](https://github.com/localstack/localstack/blob/ae31f63bb6d8254edc0c85a66e3c36cd0c7dc7b0/localstack/utils/aws/message_forwarding.py#L42) `account_id` to [send events to the target](https://github.com/localstack/localstack/blob/ae31f63bb6d8254edc0c85a66e3c36cd0c7dc7b0/localstack/utils/aws/message_forwarding.py#L31) service like sqs, sns, lambda, etc. 

- You should always refer the official AWS API docs and check how the the application is communicating to each other. 
    
    For example: 
    Firehose to s3 refer to [Offical AWS Docs](https://docs.aws.amazon.com/firehose/latest/dev/controlling-access.html#cross-account-delivery-s3).

## Test changes in CI with non-default credentials

> We regularly run the circleci test jobs of the LocalStack community repository to check the compatibility of cross-account against the changes. 

Reviewer: There are several references to "Community" still in this page, but @HarshCasper mentioned in a comment above that we should get rid of these.
```suggestion
We regularly run the circleci test jobs of the LocalStack repository to check the compatibility of cross-account against the changes. 
```

To achieve that, we have a [scheduled workflow](https://github.com/localstack/localstack/blob/master/.circleci/config.yml) on [LocalStack](https://github.com/localstack/localstack), which executes the tests with randomised account and region credentials every night at 1:00am UTC.

> To manually trigger the scheduled workflow through circleci UI(in case you have the permissions), in order to test your changes with randomised account and region credentials, you can perform the following steps: 

Reviewer: ```suggestion
To manually trigger the workflow through the webinterface of CircleCI (in case you have the permissions), in order to test your changes with randomized account and region credentials, you can perform the following steps: 
```

> - Go to the [localStack](https://app.circleci.com/pipelines/github/localstack/localstack) project repository on circleci UI.
> - Select a branch for which you want to trigger the scheduled workflow from the filters section.
> - Now click on the `Trigger pipeline` button on the right and add the following variables:
>     1. Parameter type to `string`
>     2. Name to `randomize-aws-credentials`
>     3. Value to `true`
> 
> and press the `Trigger pipeline` button to trigger the workflow.

Reviewer: ```suggestion
- Go to the [LocalStack](https://app.circleci.com/pipelines/github/localstack/localstack) project repository on CircleCI.
- Select a branch for which you want to trigger the workflow from the filters section.
- Now click on the `Trigger pipeline` button on the right and add the following variables:
    1. Parameter type to `string`
    2. Name to `randomize-aws-credentials`
    3. Value to `true`
- Press the `Trigger pipeline` button to trigger the workflow.
```
